File: 0198-house-robber/0198-house-robber.c
int rob(int* nums, int n) {
    int arr[n+1];
    if(n == 0){
        return 0;
    }
    if(n == 1){
        return nums[0];
    }
    arr[0] = nums[0];
    arr[1] = max(nums[1],nums[0]);
    for(int i = 2;i<n;i++){
        arr[i] = max(arr[i-1],nums[i]+arr[i-2]);
    }
    return arr[n-1];

}

int max(int a,int b){
    return a >= b ? a : b;
}

File: 0198-house-robber/0198-house-robber.java
class Solution {
    int[] dp;
    public int rob(int[] nums) {
        int n = nums.length;
        dp = new int[n];
        Arrays.fill(dp,-1);
        return solve(nums,n-1);
    }

    int solve(int[] nums,int i){
        if(i < 0){
            return 0;
        }
        if(dp[i] != -1) return dp[i];
        if(i == 0){
            return dp[i] =  nums[i];
        }
        int take = nums[i] + solve(nums,i-2);
        int notTake = solve(nums,i-1);
        return dp[i] = Math.max(take,notTake);
    }

}